protocols/noise/src/io.rs

```diff
@@ -20,448 +20,117 @@
-        }
-    }
-}
+use log::trace;
+use std::{cmp::min, fmt, io, pin::Pin, task::{Context, Poll}};
 
 /// A noise session to a remote.
 ///
 /// `T` is the type of the underlying I/O resource.
 pub struct NoiseOutput<T> {
-    io: T,
-    session: SnowState,
-    read_state: ReadState,
-    write_state: WriteState,
-    read_buffer: Vec<u8>,
-    write_buffer: Vec<u8>,
-    decrypt_buffer: Vec<u8>,
-    encrypt_buffer: Vec<u8>
+    io: NoiseFramed<T, snow::TransportState>,
+    recv_offset: usize,
+    recv_buffer: Vec<u8>,
+    send_buffer: Vec<u8>
 }
 
 impl<T> fmt::Debug for NoiseOutput<T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("NoiseOutput")
-            .field("read_state", &self.read_state)
-            .field("write_state", &self.write_state)
+            .field("io", &self.io)
             .finish()
     }
 }
 
 impl<T> NoiseOutput<T> {
-    fn new(io: T, session: SnowState) -> Self {
+    fn new(io: NoiseFramed<T, snow::TransportState>) -> Self {
         NoiseOutput {
             io,
-            session,
-            read_state: ReadState::Init,
-            write_state: WriteState::Init,
-            read_buffer: Vec::new(),
-            write_buffer: Vec::new(),
-            decrypt_buffer: Vec::new(),
-            encrypt_buffer: Vec::new()
+            recv_buffer: Vec::new(),
+            send_buffer: Vec::new(),
+            recv_offset: 0,
         }
     }
 }
 
-/// The various states of reading a noise session transitions through.
-#[derive(Debug)]
-enum ReadState {
-    /// initial state
-    Init,
-    /// read frame length
-    ReadLen { buf: [u8; 2], off: usize },
-    /// read encrypted frame data
-    ReadData { len: usize, off: usize },
-    /// copy decrypted frame data
-    CopyData { len: usize, off: usize },
-    /// end of file has been reached (terminal state)
-    /// The associated result signals if the EOF was unexpected or not.
-    Eof(Result<(), ()>),
-    /// decryption error (terminal state)
-    DecErr
-}
-
-/// The various states of writing a noise session transitions through.
-#[derive(Debug)]
-enum WriteState {
-    /// initial state
-    Init,
-    /// accumulate write data
-    BufferData { off: usize },
-    /// write frame length
-    WriteLen { len: usize, buf: [u8; 2], off: usize },
-    /// write out encrypted data
-    WriteData { len: usize, off: usize },
-    /// end of file has been reached (terminal state)
-    Eof,
-    /// encryption error (terminal state)
-    EncErr
-}
-
 impl<T: AsyncRead + Unpin> AsyncRead for NoiseOutput<T> {
-    fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<io::Result<usize>> {
-        let mut this = self.deref_mut();
+    fn poll_read(self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<io::Result<usize>> {
+        let this = Pin::into_inner(self);
```

Comment:
I think within `poll_read`, one could directly use `self`.

Reply:
Done in https://github.com/libp2p/rust-libp2p/pull/1658/commits/062205f741fa151f5eed0d17036efe3528fd16a2.